File: plugins/templates/AudioFilterModule.java
package com.soundscape;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import android.media.MediaPlayer;
import android.media.audiofx.Equalizer;

import java.io.IOException;

public class AudioFilterModule extends ReactContextBaseJavaModule {
  private final ReactApplicationContext reactContext;

  public AudioFilterModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @Override
  public String getName() {
    return "AudioFilterModule";
  }

  @ReactMethod
  public void filterAndPlay(String filePath, int cutoff) {
    MediaPlayer player = new MediaPlayer();
    try {
      player.setDataSource(filePath);
      player.setLooping(true);
      player.prepare();
    } catch (IOException e) {
      e.printStackTrace();
      return;
    }

    // Create Equalizer and attach to player
    Equalizer eq = new Equalizer(0, player.getAudioSessionId());
    eq.setEnabled(true);
    short bands      = eq.getNumberOfBands();
    short[] range    = eq.getBandLevelRange();

    for (short i = 0; i < bands; i++) {
      int freq = eq.getCenterFreq(i) / 1000; // in Hz
      if (freq > cutoff) {
        eq.setBandLevel(i, range[0]); // silence above cutoff
      } else {
        eq.setBandLevel(i, range[1]); // full volume below
      }
    }

    player.start();
  }
}

File: android/app/src/main/java/com/tai/soundscape/AudioFilterModule.kt
// android/app/src/main/java/com/tai/soundscape/AudioFilterModule.kt
package com.tai.soundscape

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import android.media.audiofx.Equalizer

class AudioFilterModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private val eq: Equalizer = Equalizer(0, 0).apply {
    enabled = true
    // initialize all bands at unity (0 dB)
    for (i in 0 until numberOfBands) setBandLevel(i.toShort(), 0)
  }

  override fun getName(): String = "AudioFilterModule"

  /**
   * freq: cutoff in Hz
   */
  @ReactMethod
  fun setLowPassFrequency(freq: Float) {
    // Mute bands above cutoff, leave below at 0dB
    for (i in 0 until eq.numberOfBands) {
      val centerFreq = eq.getCenterFreq(i.toShort()) / 1000
      val level = if (centerFreq <= freq) 0 else -1500  // -15 dB
      eq.setBandLevel(i.toShort(), level.toShort())
    }
  }
}
